eth_account/_utils/structured_data/hashing.py

```diff
@@ -0,0 +1,319 @@
+
+    result = ''.join(
+        [
+            encode_struct(struct_name, types[struct_name])
+            for struct_name in sorted_deps
+        ]
+    )
+    return result
+
+
+def hash_struct_type(primary_type, types):
+    return keccak(text=encode_type(primary_type, types))
+
+
+def is_valid_abi_type(type_name):
+    """
+    This function is used to make sure that the ``type_name`` is a valid ABI Type.
+
+    Please note that this is a temporary function and should be replaced by the corresponding
+    ABI function, once the following issue has been resolved.
+    https://github.com/ethereum/eth-abi/issues/125
+    """
+    valid_abi_types = {"address", "bool", "bytes", "int", "string", "uint"}
+    is_bytesN = type_name.startswith("bytes") and 1 <= int(type_name[5:]) <= 32
+    is_intN = (
+        type_name.startswith("int") and
+        8 <= int(type_name[3:]) <= 256 and
+        int(type_name[3:]) % 8 == 0
+    )
+    is_uintN = (
+        type_name.startswith("uint") and
+        8 <= int(type_name[4:]) <= 256 and
+        int(type_name[4:]) % 8 == 0
+    )
+
+    if type_name in valid_abi_types:
+        return True
+    elif is_bytesN:
+        # bytes1 to bytes32
+        return True
+    elif is_intN:
+        # int8 to int256
+        return True
+    elif is_uintN:
+        # uint8 to uint256
+        return True
+
+    return False
+
+
+def is_array_type(type):
+    # Identify if type such as "person[]" or "person[2]" is an array
+    abi_type = parse(type)
+    return abi_type.is_array
+
+
+@to_tuple
+def get_depths_and_dimensions(data, depth):
+    """
+    Yields 2-length tuples of depth and dimension of each element at that depth
+    """
+    if not isinstance(data, list) and not isinstance(data, tuple):
+        # Not checking for Iterable instance, because even Dictionaries and strings
+        # are considered as iterables, but that's not what we want the condition to be.
+        return ()
+
+    yield depth, len(data)
+
+    for item in data:
+        # iterating over all 1 dimension less sub-data items
+        yield from get_depths_and_dimensions(item, depth + 1)
+
+
+def get_array_dimensions(data):
+    """
+    Given an array type data item, check that it is an array and
+    return the dimensions as a tuple.
+    Ex: get_array_dimensions([[1, 2, 3], [4, 5, 6]]) returns (2, 3)
+    """
+    depths_and_dimensions = get_depths_and_dimensions(data, 0)
+    # re-form as a dictionary with `depth` as key, and all of the dimensions found at that depth.
+    grouped_by_depth = {
+        depth: tuple(tuple(zip(*group))[1])
+        for depth, group in groupby(depths_and_dimensions, itemgetter(0))
+    }
+
+    # validate that there is only one dimension for any given depth.
+    invalid_depths = tuple(
+        depth
+        for depth, dimensions in grouped_by_depth.items()
+        if len(set(dimensions)) != 1
+    )
+    if invalid_depths:
+        raise ValidationError("Not a valid array or incomplete array")
```

Comment:
I think we should do the extra work to construct an informative error message here that includes each of the invalid depths and what dimensions were found at that depth.  Otherwise a user confronted with this error message is unlikely to have any idea what went wrong without a deep understanding of the signing spec which I think is not likely.